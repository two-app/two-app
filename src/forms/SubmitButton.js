import React from "react";
import PropTypes from "prop-types";
import {StyleSheet, Text, TouchableOpacity} from "react-native";
import Colors from "../Colors";

const SubmitButton = ({text, onSubmit, disabled}) => (
    <TouchableOpacity style={[styles.submitButton, disabled ? styles.disabledSubmitButton : undefined]}
                      onPress={onSubmit}
                      disabled={disabled}>
        <Text style={styles.submitText}>{text}</Text>
    </TouchableOpacity>
);

SubmitButton.propTypes = {
    text: PropTypes.string.isRequired,
    onSubmit: PropTypes.func.isRequired,
    disabled: PropTypes.bool
};

SubmitButton.defaultProps = {
    disabled: false
};

const styles = StyleSheet.create({
    submitButton: {
        justifyContent: "center",
        alignItems: "center",
        backgroundColor: Colors.SALMON,
        padding: 20,
        marginTop: 25,
        marginBottom: 25
    },
    disabledSubmitButton: {
        backgroundColor: Colors.FADED
    },
    submitText: {
        color: "white",
        fontWeight: "bold"
    }
});

export default SubmitButton;