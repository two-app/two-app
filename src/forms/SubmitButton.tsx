import {useState} from 'react';
import {
  ActivityIndicator,
  Text,
  TouchableWithoutFeedbackProps,
  View,
  ViewStyle,
} from 'react-native';
import {TouchableWithoutFeedback} from 'react-native';
import Colors from '../Colors';

type BaseButtonProps = TouchableWithoutFeedbackProps & {
  loading?: boolean;
  loadingColor?: string;
  viewStyle?: ViewStyle;
};

type StyledButtonProps = BaseButtonProps & {
  color: string;
  pressedColor: string;
};

type XButtonProps = BaseButtonProps & {
  children: string;
};

export const PrimaryButton = (props: XButtonProps) => (
  <StyledButton
    {...props}
    pressedColor={Colors.DARK_SALMON}
    color={Colors.SALMON}>
    <Text
      style={{
        color: 'white',
        fontWeight: '600',
        fontFamily: 'Montserrat-SemiBold',
      }}>
      {props.children}
    </Text>
  </StyledButton>
);

export const StyledButton = (props: StyledButtonProps) => {
  const [pressed, setPressed] = useState(false);
  return (
    <BaseButton
      onPressIn={() => setPressed(true)}
      onPressOut={() => setPressed(false)}
      viewStyle={{
        height: 45,
        backgroundColor: pressed ? props.pressedColor : props.color,
        justifyContent: 'center',
        alignItems: 'center',
      }}
      {...props}
    />
  );
};

export const BaseButton = (props: BaseButtonProps) => {
  const loadingColor = props.loadingColor ?? 'white';

  return (
    <View style={props.style}>
      <TouchableWithoutFeedback {...props}>
        <View style={[{flexDirection: 'row'}, props.viewStyle]}>
          {props.children}
          {props.loading && <ButtonLoadingIndicator color={loadingColor} />}
        </View>
      </TouchableWithoutFeedback>
    </View>
  );
};

const ButtonLoadingIndicator = ({color}: {color: string}) => (
  <ActivityIndicator color={color} style={{marginLeft: 10}} size="small" />
);
