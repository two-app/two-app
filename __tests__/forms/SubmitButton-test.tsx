import 'react-native';
import React from 'react';
// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import SubmitButton from "../../src/forms/SubmitButton";
import {shallow} from "enzyme";
import {ShallowWrapper} from "enzyme";

test('should maintain snapshot', () => expect(
    renderer.create(<SubmitButton text={"Submit"} onSubmit={() => null}/>)
).toMatchSnapshot());

let onSubmitFn: jest.Mock;
let wrapper: ShallowWrapper;

beforeEach(() => {
    onSubmitFn = jest.fn();
    wrapper = shallow(<SubmitButton onSubmit={onSubmitFn} text="Test Submit"/>);
});

test('should callback onSubmit when pressed', () => {
    wrapper.find("TouchableOpacity").prop<() => void>("onPress")();
    expect(onSubmitFn).toHaveBeenCalled();
});

test('should display given text', () => expect(
    wrapper.find("Text").render().text()
).toEqual("Test Submit"));

test('should not be disabled by default', () => expect(
    wrapper.find("TouchableOpacity").prop("disabled")
).toBe(false));