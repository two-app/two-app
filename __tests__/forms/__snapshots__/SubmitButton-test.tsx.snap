// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should maintain snapshot 1`] = `
<TouchableOpacity
  disabled={false}
  onPress={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#EE7674",
        "justifyContent": "center",
        "marginBottom": 25,
        "marginTop": 25,
        "padding": 20,
      },
      undefined,
    ]
  }
>
  <Text
    style={
      Object {
        "color": "white",
        "fontWeight": "bold",
      }
    }
  >
    Submit
  </Text>
</TouchableOpacity>
`;
